src/representation_theory/basic.lean

```diff
@@ -107,4 +128,121 @@ theorem char_conj (g : G) (h : G) : (character k G V) (h * g * h⁻¹) = (charac
+                          smul_apply, map_smulₛₗ, ring_hom.id_apply]}}
+
+instance dual_distrib_mul_action : distrib_mul_action G (module.dual k V)  :=
+{ smul := λ g f, f ∘ₗ ↑(as_group_hom k G V g)⁻¹,
+  one_smul := λ f, by {ext, simp only [map_one, one_inv, units.coe_one, coe_comp, function.comp_app,
+                      one_apply]},
+  mul_smul := λ g h f, by {ext, simp only [map_mul, mul_inv_rev, units.coe_mul, coe_comp,
+                          function.comp_app, linear_map.mul_apply]},
+  smul_add := λ g f₁ f₂, by simp only [add_comp],
+  smul_zero := λ g, by simp only [zero_comp]}
+
+@[simp]
+lemma dual_smul_def (g : G) (f : module.dual k V) : g • f = f ∘ₗ ↑(as_group_hom k G V g)⁻¹ := rfl
+
+instance dual_smul_comm_class : smul_comm_class G k (module.dual k V) :=
+{smul_comm := λ g r f, by {ext, simp only [dual_smul_def, coe_comp, function.comp_app, smul_apply]}}
+
+end linear_hom
+
+section tensor
+
+open_locale tensor_product
+
+variables (k G V W : Type*) [comm_semiring k] [monoid G] [add_comm_monoid V] [add_comm_monoid W]
+variables [module k V] [distrib_mul_action G V] [smul_comm_class G k V]
+variables [module k W] [distrib_mul_action G W] [smul_comm_class G k W]
+
+/--
+Given two `k`-linear representations `V` and `W`, their tensor product over `k` is itself a
+representation, characterized by `g • (v ⊗ₜ w) = (g • v) ⊗ₜ (g • w)` for `g : G`, `v : V`, `w : W`.
+-/
+@[derive [add_comm_monoid, module k, inhabited]]
+def tensor_rep : Type* := (V ⊗[k] W)
+
+notation V ` ⊗[` k `,` G `] ` W := tensor_rep k V W
+
+variables {k G V W}
+
+/--
+The tensor product representation `V ⊗[k,G] W` is equivalent to `V ⊗[k] W` as a `k`-module
+-/
+@[simp]
+def of_tensor_rep : (V ⊗[k,G] W) ≃ₗ[k] (V ⊗[k] W) :=
+  ⟨id, λ _ _, rfl, λ _ _, rfl, id, λ _, rfl, λ _, rfl⟩
+
+/--
+The tensor product representation `V ⊗[k,G] W` is equivalent to `V ⊗[k] W` as a `k`-module
+-/
+@[simp]
+def to_tensor_rep : (V ⊗[k] W) ≃ₗ[k] (V ⊗[k,G] W) := (of_tensor_rep).symm
+
+notation v ` ⊗ₜ[` k `,` G `] ` w := to_tensor_rep (v ⊗ₜ[k] w)
+
+instance tensor_rep_distrib_mul_action : distrib_mul_action G (V ⊗[k,G] W) :=
+{ smul := λ g, tensor_product.map
+          (to_module_End k V g) (to_module_End k W g),
+  one_smul := λ x, by simp only [monoid_hom.map_one, tensor_product.map_one, one_apply],
+  mul_smul := λ g h x, by simp [monoid_hom.map_mul, mul_eq_comp, tensor_product.map_comp],
+  smul_add := λ g x₁ x₂, by simp only [map_add],
+  smul_zero := λ g, by simp only [map_zero]}
+
+@[simp]
+lemma tensor_rep_def (g : G) (x : V ⊗[k,G] W) :
+  g • x = to_tensor_rep (tensor_product.map
+    (to_module_End k V g) (to_module_End k W g) (of_tensor_rep x)) := rfl
+
+instance tensor_rep_smul_comm_class : smul_comm_class G k (V ⊗[k,G] W) :=
+{smul_comm := λ g r x, by simp}
+
+@[simp]
+lemma tensor_rep_to_linear_map (g : G) : to_linear_map k (V ⊗[k,G] W) g = tensor_product.map
+          (to_linear_map k V g) (to_linear_map k W g) :=
```

Comment:
This one has the wrong type too, it should be composed with the appropriate to / `of| functions on the left and right